File: JAVA/Day 8/smallest_change_pali.java
/*
Mr Shravan is given a word W, and W is a palindrome. Mr Shravan is 
allowed to replace only one letter in W, with any lowercase alphabet, 
such that the word W is not a palindrome again and which is the 
lexicographically smallest among all the possible options.

Examples of Lexicographical order are:
	- aaa is smaller than aab
	- abb is smaller than abc.

Return the lexicographically smallest string among all the possible options.

Input Format:
-------------
A String P, palindrome.

Output Format:
--------------
A String result.


Sample Input-1:
---------------
abcdcba

Sample Output-1:
----------------
aacdcba


Sample Input-2:
---------------
aaaa

Sample Output-2:
----------------
aaab


Sample Input-3:
---------------
aaabaaa

Sample Output-3:
----------------
aaabaab

*/

import java.util.*;
class Test{
  public static void main(){
		Scanner sc = new Scanner(System.in);
		String s = sc.nextLine();
		for(int i=0;i<s.length()/2;i++){
			if(s.charAt(i)!='a'){
				s.charAt(i) = 'a';
				System.out.println(s);
				System.exit(0);
			}
		}
		s[s.length()-1] = 'b';
		System.out.println(s);
	}
}
		

File: CPP/Day 9/smallest_change_pali.cpp
/*
/*
Mr Shravan is given a word W, and W cosists of lowercase alphabets and '#'. 
Mr Shravan is allowed to replace the '#' with any one lowercase alphabet, 
such that the word W is a palindrome and it has to be the lexicographically
smallest among all the possible options.

Your task is to help Mr Sharavan to return the lexicographically smallest 
palindrome string among all the possible options. 
OR "invlaid" if it is not possible.

Input Format:
-------------
A String W, consists of lowercase letters and #.

Output Format:
--------------
A String result.


Sample Input-1:
---------------
r#d#v##er

Sample Output-1:
----------------
redavader


Sample Input-2:
---------------
r#d#v#cer

Sample Output-2:
----------------
invalid
*/

#include <bits/stdc++.h>
using namespace std;

string palindrome(string s,int n){
    int start=0;
    int end=n-1;
    while(start<=end){
        if(s[start]=='#' && s[end]=='#'){
            s[start]='a';
            s[end]='a';
            start++;
            end--;
        }
        else if(s[start]=='#' && isalpha(s[end])){
            s[start] = s[end];
            start++;
            end--;
        }
        else if(s[end]=='#' && isalpha(s[start])){
            s[end] = s[start];
            start++;
            end--;
        }
        else if(s[start]==s[end]){
            start++;
            end--;
        }
        else{
            return "invalid";
        }
    }
    return s;
}
int main(){
    string s;
    cin>>s;
    cout<<palindrome(s,s.size())<<endl;
}
